demo: panic when parsing connection string

**Describe the problem**

When running `cockroach demo`, I used the `\connect` command to attempt to connect to the system tenant, with the connection string output in the console.

I hit the following panic:

```console
demo@127.0.0.1:26257/? ?> \connect postgresql://demo:demo2559099@127.0.0.1:26258/defaultdb?sslmode=require&sslrootcert=%2Ftmp%2Fdemo3599721519%2Fca.crt
*
* ERROR: a panic has occurred!
* runtime error: invalid memory address or nil pointer dereference
* (1) attached stack trace
*   -- stack trace:
*   | runtime.sigpanic
*   |   GOROOT/src/runtime/signal_unix.go:735
*   | github.com/cockroachdb/cockroach/pkg/cli/clienturl.(*urlParser).SetHost
*   |   github.com/cockroachdb/cockroach/pkg/cli/clienturl/client_url.go:180
*   | github.com/cockroachdb/cockroach/pkg/security/clientsecopts.AnalyzeClientURL
*   |   github.com/cockroachdb/cockroach/pkg/security/clientsecopts/analyze_url_flag.go:106
*   | github.com/cockroachdb/cockroach/pkg/cli/clienturl.(*urlParser).Set
*   |   github.com/cockroachdb/cockroach/pkg/cli/clienturl/client_url.go:122
*   | github.com/cockroachdb/cockroach/pkg/cli/clienturl.MakeURLParserFn.func1
*   |   github.com/cockroachdb/cockroach/pkg/cli/clienturl/client_url.go:253
*   | github.com/cockroachdb/cockroach/pkg/cli/clisqlshell.(*cliState).handleConnectInternal
*   |   github.com/cockroachdb/cockroach/pkg/cli/clisqlshell/sql.go:1435
*   | github.com/cockroachdb/cockroach/pkg/cli/clisqlshell.(*cliState).handleConnect
*   |   github.com/cockroachdb/cockroach/pkg/cli/clisqlshell/sql.go:1414
*   | github.com/cockroachdb/cockroach/pkg/cli/clisqlshell.(*cliState).doHandleCliCmd
*   |   github.com/cockroachdb/cockroach/pkg/cli/clisqlshell/sql.go:1297
*   | github.com/cockroachdb/cockroach/pkg/cli/clisqlshell.(*cliState).doRunShell
*   |   github.com/cockroachdb/cockroach/pkg/cli/clisqlshell/sql.go:1981
*   | github.com/cockroachdb/cockroach/pkg/cli/clisqlshell.(*cliState).RunInteractive
*   |   github.com/cockroachdb/cockroach/pkg/cli/clisqlshell/sql.go:1930
*   | github.com/cockroachdb/cockroach/pkg/cli/clisqlcfg.(*Context).Run
*   |   github.com/cockroachdb/cockroach/pkg/cli/clisqlcfg/context.go:222
*   | github.com/cockroachdb/cockroach/pkg/cli.runDemo
*   |   github.com/cockroachdb/cockroach/pkg/cli/demo.go:358
*   | github.com/cockroachdb/cockroach/pkg/cli.init.5.func1
*   |   github.com/cockroachdb/cockroach/pkg/cli/demo.go:58
*   | github.com/cockroachdb/cockroach/pkg/cli/clierrorplus.MaybeDecorateError.func1
*   |   github.com/cockroachdb/cockroach/pkg/cli/clierrorplus/decorate_error.go:67
*   | github.com/spf13/cobra.(*Command).execute
*   |   github.com/spf13/cobra/external/com_github_spf13_cobra/command.go:856
*   | github.com/spf13/cobra.(*Command).ExecuteC
*   |   github.com/spf13/cobra/external/com_github_spf13_cobra/command.go:974
*   | github.com/spf13/cobra.(*Command).Execute
*   |   github.com/spf13/cobra/external/com_github_spf13_cobra/command.go:902
*   | github.com/cockroachdb/cockroach/pkg/cli.Run
*   |   github.com/cockroachdb/cockroach/pkg/cli/cli.go:298
*   | github.com/cockroachdb/cockroach/pkg/cli.doMain
*   |   github.com/cockroachdb/cockroach/pkg/cli/cli.go:141
*   | github.com/cockroachdb/cockroach/pkg/cli.Main
*   |   github.com/cockroachdb/cockroach/pkg/cli/cli.go:66
*   | main.main
*   |   main/pkg/cmd/cockroach-short/main.go:26
*   | runtime.main
*   |   GOROOT/src/runtime/proc.go:255
*   | runtime.goexit
*   |   GOROOT/src/runtime/asm_amd64.s:1581
* Wraps: (2) runtime error: invalid memory address or nil pointer dereference
* Error types: (1) *withstack.withStack (2) runtime.errorString
*
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x409e6f6]
```

**To Reproduce**

Run `cockroach demo`, then `\connect $URL`.

This reliably reproduces on on `master` @ 2e1c19b26700d05ca1218a1dc3c6bf1d5ec11cde.

**Expected behavior**

No panic. If I, as the user, enter a command, even if I'm holding it wrong (i.e. PEBKAC), I should get an error rather than crashing the process.

--- a/pkg/cli/clienturl/client_url.go
+++ b/pkg/cli/clienturl/client_url.go
@@ -177,13 +177,21 @@ func (u *urlParser) SetUser(user string) {
 // SetHost implements the clientsecopts.CLIFlagInterfaceForClientURL interface.
 func (u *urlParser) SetHost(host string) {
 	u.clientOpts.ServerHost = host
-	u.fl.Lookup(cliflags.ClientHost.Name).Changed = true
+	// Also update the flag, so that subsequent CLI parameters
+	// pick up the change. NB: 'cockroach demo' does not have a --host flag.
+	if f := u.fl.Lookup(cliflags.ClientHost.Name); f != nil {
+		f.Changed = true
+	}
 }
 
 // SetPort implements the clientsecopts.CLIFlagInterfaceForClientURL interface.
 func (u *urlParser) SetPort(port string) {
 	u.clientOpts.ServerPort = port
-	u.fl.Lookup(cliflags.ClientPort.Name).Changed = true
+	// Also update the flag, so that subsequent CLI parameters
+	// pick up the change. NB: 'cockroach demo' does not have a --host flag.
+	if f := u.fl.Lookup(cliflags.ClientPort.Name); f != nil {
+		f.Changed = true
+	}
 }
 
 // SetDatabase implements the clientsecopts.CLIFlagInterfaceForClientURL interface.
@@ -208,19 +216,27 @@ func (u *urlParser) SetDatabase(db string) {
 // CertsDirFlag implements the clientsecopts.CLIFlagInterfaceForClientURL interface.
 func (u *urlParser) CertsDirFlag() (bool, string) {
 	flCertsDir := u.fl.Lookup(cliflags.CertsDir.Name)
+	if flCertsDir == nil {
+		return false, ""
+	}
 	return flCertsDir.Changed, flCertsDir.Value.String()
 }
 
 // SetCertsDir implements the clientsecopts.CLIFlagInterfaceForClientURL interface.
 func (u *urlParser) SetCertsDir(certsDir string) {
 	flCertsDir := u.fl.Lookup(cliflags.CertsDir.Name)
-	_ = flCertsDir.Value.Set(certsDir)
-	flCertsDir.Changed = true
+	if flCertsDir != nil {
+		_ = flCertsDir.Value.Set(certsDir)
+		flCertsDir.Changed = true
+	}
 }
 
 // InsecureFlag implements the clientsecopts.CLIFlagInterfaceForClientURL interface.
 func (u *urlParser) InsecureFlag() (bool, bool) {
 	flInsecure := u.fl.Lookup(cliflags.ClientInsecure.Name)
+	if flInsecure == nil {
+		return false, false
+	}
 	vb, _ := strconv.ParseBool(flInsecure.Value.String())
 	return flInsecure.Changed, vb
 }
@@ -228,7 +244,9 @@ func (u *urlParser) InsecureFlag() (bool, bool) {
 // SetInsecure implements the clientsecopts.CLIFlagInterfaceForClientURL interface.
 func (u *urlParser) SetInsecure(insecure bool) {
 	flInsecure := u.fl.Lookup(cliflags.ClientInsecure.Name)
-	_ = flInsecure.Value.Set(strconv.FormatBool(insecure))
+	if flInsecure != nil {
+		_ = flInsecure.Value.Set(strconv.FormatBool(insecure))
+	}
 }
 
 // NewStrictTLSConfigurationError implements the clientsecopts.CLIFlagInterfaceForClientURL interface.

--- a/pkg/cli/interactive_tests/test_demo.tcl
+++ b/pkg/cli/interactive_tests/test_demo.tcl
@@ -196,6 +196,14 @@ eexpect "sslmode=require"
 eexpect "sslrootcert="
 eexpect "defaultdb>"
 
+end_test
+
+start_test "Check that invalid URL is rejected"
+# Regression test for 83598
+send "\\connect postgresql://foo:123/\r"
+eexpect "using new connection URL"
+eexpect "failed to connect"
+eexpect " ?>"
 send_eof
 eexpect eof
 

